### PR TITLE
HUS-2225 multimedia alt and title transfer to Shopware

### PR DESCRIPTION
**Pull request by strix-ladamczuk**

### CODE REVIEW

**Review by rklak — APPROVED** [review]

File: src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+                    $altTranslations[$shopwareLanguage->getIso()] ?? null,
+                    $titleTranslations[$shopwareLanguage->getIso()] ?? null,
+                    $shopwareLanguage->getId()
+                );
+            }
+        }
+        $shopware6Media->updateTranslated($ergoTranslations);
+
+        if ($shopware6Media->isModified()) {
+            $this->logger->info(
```
Comment (kardi31) [inline src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php:101]: Please set it to `debug`. We dont want to log it every time the export is run

File: src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            );
+        }
+    }
+
+    /**
+     * @throws GuzzleException
+     */
+    private function updateTranslations(Shopware6Channel $channel, Shopware6Media $shopware6Media): void
+    {
+        if (!empty($shopware6Media->getTranslations())) {
```
Comment (kardi31) [inline src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php:126]: what if translations are removed from Ergonode? We should check `isModified` instead

File: src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        $ergoTranslations = [];
+        foreach ($channel->getLanguages() as $channelLanguage) {
+            if (!in_array($channelLanguage->getCode(), $languages)) {
+                continue;
+            }
+            if ($this->languageRepository->exists($channel->getId(), $channelLanguage->getCode())) {
+                $shopwareLanguage = $this->languageRepository->load($channel->getId(), $channelLanguage->getCode());
+                Assert::notNull(
+                    $shopwareLanguage,
+                    sprintf('Expected a value other than null for product lang  %s', $channelLanguage->getCode())
```
Comment (kardi31) [inline src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php:88]: Can we change the part `product lang` to `media update lang` or similar. It's for exception to be more specific so we know which part of app it's thrown

File: src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    /**
+     * @throws GuzzleException
+     */
+    private function ensureMediaTranslations(
+        Shopware6Channel $channel,
+        AbstractMultimedia $multimedia,
+        string $shopwareId
+    ): void {
+        //apply translations from shopware
+        $shopware6Media = new Shopware6Media($shopwareId, null);
```
Comment (kardi31) [inline src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php:73]: I think we should check in the repository for shopware6 media whether this shopwareid is stored in our database

File: src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    private function getMediaTranslations(Shopware6Channel $channel, string $shopwareMediaId): array
+    {
+        $query = new Shopware6QueryBuilder();
+        $query->limit(1000);
+
+        $action = new GetMediaTranslations($query, $shopwareMediaId);
+
+        try {
+            return $this->connector->execute($channel, $action);
+        } catch (ClientException $exception) {
```
Comment (kardi31) [inline src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php:145]: why we ignore this exception?

File: src/Infrastructure/Connector/Action/Media/GetMedia.php [hunk 6]
```diff
@@ -44,7 +44,10 @@ public function parseContent(?string $content): ?Shopware6Media
 
         $data = json_decode($content, true, 512, JSON_THROW_ON_ERROR);
 
-        return new Shopware6Media($data['data']['id'], $data['data']['attributes']['fileName']);
+        return new Shopware6Media(
+            $data['data']['id'],
+            $data['data']['attributes']['fileName'] ?: null
```
Comment (lajtof) [inline src/Infrastructure/Connector/Action/Media/GetMedia.php:49]: are you sure `$data['data']['attributes']['fileName']` always exist?

`$data['data']['attributes']['fileName'] ?? null` would be safer but I'm not sure what's the purpose of this change ;)

File: src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php (hunk 2 above)
Comment (strix-ladamczuk) [reply src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php:126]: It's already done in one update action - since shopware API has difficulties with  removal records completely - data market as deleted are marked as empty on check - see loop on Media Model file.

File: src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        AbstractMultimedia $multimedia,
+        string $shopwareId
+    ): void {
+        //apply translations from shopware
+        $shopware6Media = new Shopware6Media($shopwareId, null);
+        $shopware6Media->setTranslations($this->getMediaTranslations($channel, $shopwareId));
+        //apply translations from ergo
+        $altTranslations = $multimedia->getAlt()->getTranslations();
+        $titleTranslations = $multimedia->getTitle()->getTranslations();
+        $languages = array_unique(array_merge(array_keys($altTranslations), array_keys($titleTranslations)));
```
Comment (kardi31) [inline src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php:78]: This logic looks like a potential risk. Are we should all those array operations always return correct data? I would much rather loop through existing translations from Shopware, check if data has changed in Ergonode and apply changes

**Review by kardi31 — CHANGES_REQUESTED** [review]
Some feedback

File: src/Infrastructure/Model/Shopware6Media.php [hunk 8]
```diff
@@ -37,6 +47,92 @@ public function setFileName(string $fileName): void
...
+            if (!in_array($language, $processedLanguages) && !$this->isTranslationEmpty($translation)) {
+                $this->translations[$language] = new Shopware6MediaTranslation(
+                    $translation->getId(),
+                    null,
+                    null,
+                    $translation->getLanguageId()
+                );
+                $this->setModified();
+            }
+        }
```
Comment (lajtof) [inline src/Infrastructure/Model/Shopware6Media.php:126]: can't it be simplified a bit?

I mean we could implement early exit/continue, or omit checking if translation is changed, and probably more

File: src/Infrastructure/Connector/Action/Media/GetMedia.php (hunk 6 above)
Comment (strix-ladamczuk) [reply src/Infrastructure/Connector/Action/Media/GetMedia.php:49]: nice catch thank You

File: src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php (hunk 7 above)
Comment (strix-ladamczuk) [reply src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php:78]: In here we are "converting" / "collecting" ergonode data to later apply on Media Model.
The check that you are mention of is in the Model itself - based on this we are able to get isModified flag

File: src/Infrastructure/Model/Shopware6Media.php (hunk 8 above)
Comment (strix-ladamczuk) [reply src/Infrastructure/Model/Shopware6Media.php:126]: extracted one method to apply more visibility.
purpose of this method is to calculate if data are modified, do not think I can't simplify it more.

File: src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php (hunk 3 above)
Comment (strix-ladamczuk) [reply src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php:88]: Done

File: src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php (hunk 1 above)
Comment (strix-ladamczuk) [reply src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php:101]: Done

File: src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php (hunk 5 above)
Comment (strix-ladamczuk) [reply src/Infrastructure/Processor/Process/MediaTranslationShopware6ExportProcess.php:145]: thank, you - fixed - passed to export errors

File: src/Infrastructure/Model/Shopware6Media.php [hunk 9]
```diff
@@ -37,6 +47,100 @@ public function setFileName(string $fileName): void
...
+                continue;
+            }
+            $this->translations[$translation->getLanguageId()] = new Shopware6MediaTranslation(
+                null,
+                $translation->getAlt() ?? null,
+                $translation->getTitle() ?? null,
+                $translation->getLanguageId()
+            );
+            $this->setModified();
+        }
```
Comment (lajtof) [inline src/Infrastructure/Model/Shopware6Media.php:115]:
```suggestion
        foreach ($translations as $translation) {
            $processedLanguages[] = $translation->getLanguageId();
            if (array_key_exists($translation->getLanguageId(), $this->translations)) {
                continue;
            }
            
            if ($this->isTranslationChanged($translation, $this->translations[$translation->getLanguageId()])) {
                $id = $this->translations[$translation->getLanguageId()]->getId();
            }
            
            $this->translations[$translation->getLanguageId()] = new Shopware6MediaTranslation(
                $id ?? null,
                $translation->getAlt() ?? null,
                $translation->getTitle() ?? null,
                $translation->getLanguageId()
            );
            $this->setModified();
        }
```

**Review by kardi31 — APPROVED** [review]

File: src/Infrastructure/Model/Shopware6Media.php (hunk 9 above)
Comment (strix-ladamczuk) [reply src/Infrastructure/Model/Shopware6Media.php:115]: @filip we cannot continue until check if isTranslationChanged (found one) we cannot check on not found translations.

Comment (strix-ladamczuk) [reply src/Infrastructure/Model/Shopware6Media.php]: condition reverted for better visibility

**Review by lajtof — APPROVED** [review]
